### PR TITLE
Adding Revision Support Feature 

### PR DESCRIPTION
**Pull request by CrowleyRajapakse**
## Issues

wso2/product-apim#9406

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java [hunk 1]
```diff
@@ -1746,4 +1746,277 @@ public void setEndpointsToBeRemoved(API api, GatewayAPIDTO gatewayAPIDTO) {
...
+     * @param environment               - Gateway environment
+     * @param api                       - The API to be published
+     * @param builder                   - The template builder
+     * @param tenantDomain              - Tenant Domain of the publisher
+     * @param isGatewayDefinedAsALabel  - Whether the environment is from a label or not. If it is from a label,
+     *                                  directly publishing to gateway will not happen
+     * @param failedGatewaysMap         - This map will be updated with the gateway if the publishing failed
+     * @return failedEnvironmentsMap
+     */
+    private Map<String, String> deployAPIRevisionToGatewayEnvironment(Environment environment, API api,
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java:1832]: seems like some formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java [hunk 2]
```diff
@@ -1746,4 +1746,277 @@ public void setEndpointsToBeRemoved(API api, GatewayAPIDTO gatewayAPIDTO) {
...
+                Environment environment = getEnvironmentFromLabel(label);
+                if (debugEnabled) {
+                    log.debug("API with " + api.getId() + " is removing from the label " + label);
+                }
+                failedEnvironmentsMap = removeAPIRevisionFromGatewayEnvironment(api, tenantDomain, environment,
+                        true, removedGateways, failedEnvironmentsMap);
+            }
+        }
+
+        DeployAPIInGatewayEvent
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java:1954]: is this formatted?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 3]
```diff
@@ -1823,58 +1760,57 @@ public API updateAPI(API api, API existingAPI) throws APIManagementException, Fa
...
-            if (gatewayExists && isAPIPublished && !oldApi.getUriTemplates().equals(api.getUriTemplates())) {
-                Set<URITemplate> resourceVerbs = api.getUriTemplates();
-                if (resourceVerbs != null) {
-                    invalidateResourceCache(api.getContext(), api.getId().getVersion(), resourceVerbs);
-                    if (log.isDebugEnabled()) {
-                        log.debug("Calling invalidation cache");
-                    }
-                }
-            }
+//            APIManagerConfiguration config = ServiceReferenceHolder.getInstance().
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:1763]: lets remove the commented out part

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 4]
```diff
@@ -2672,20 +2573,19 @@ public void changeAPIStatus(API api, APIStatus status, String userId, boolean up
...
-                            failedGateways = publishToGateway(api);
-                            //Sending Notifications to existing subscribers
-                            if (APIConstants.PUBLISHED.equals(newStatus)) {
-                                sendEmailNotification(api);
-                            }
-                        } else { // API Status : RETIRED or CREATED
-                            failedGateways = removeFromGateway(api);
-                        }
-                    }
+//                    if (APIConstants.API_GATEWAY_TYPE_SYNAPSE.equalsIgnoreCase(gatewayType)) {
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2576]: remove

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 5]
```diff
@@ -2732,14 +2631,14 @@ public void changeAPIStatus(API api, APIStatus status, String userId, boolean up
...
-                            failedGateways = publishToGateway(api);
-                        } else { // API Status : RETIRED or CREATED
-                            failedGateways = removeFromGateway(api);
-                        }
-                    }
+//                    if (APIConstants.API_GATEWAY_TYPE_SYNAPSE.equalsIgnoreCase(gatewayType)) {
+//                        if (APIConstants.PUBLISHED.equals(newStatus) || APIConstants.DEPRECATED.equals(newStatus)
+//                            || APIConstants.BLOCKED.equals(newStatus) || APIConstants.PROTOTYPED.equals(newStatus)) {
+//                            failedGateways = publishToGateway(api);
+//                        } else { // API Status : RETIRED or CREATED
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:2638]: remove

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 6]
```diff
@@ -446,7 +447,62 @@ public API getAPI(APIIdentifier identifier) throws APIManagementException {
 
             API api = APIUtil.getAPIForPublishing(apiArtifact, registry);
             APIUtil.updateAPIProductDependencies(api, registry);
+            api.setSwaggerDefinition(getOpenAPIDefinition(identifier));
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:450]: Is there any place where this getAPI() method getting called? Ideally this method should not get executed.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java [hunk 7]
```diff
@@ -1031,7 +1087,12 @@ public boolean deleteWsdl(String wsdlId) throws APIManagementException {
      */
     @Override
     public ResourceFile getWSDL(APIIdentifier apiId) throws APIManagementException {
-        String wsdlResourcePath = APIConstants.API_WSDL_RESOURCE_LOCATION +
+        String apiPath = APIUtil.getAPIPath(apiId);
+        int prependIndex = apiPath.indexOf(apiId.getVersion()) + apiId.getVersion().length();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:1091]: same here getWSDL() should not be called

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 6 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:450]: no, this change is coming from old code, no uses for Revisions now

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java (hunk 7 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/AbstractAPIManager.java:1091]: no, this change is coming from old code, no uses for Revisions now

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 8]
```diff
@@ -276,6 +276,168 @@ public PublisherAPI addAPI(Organization org, PublisherAPI publisherAPI) throws A
...
+    @Override
+    public void deleteAPIRevision(Organization org, String apiUUID, int revisionId)
+            throws APIPersistenceException {
+        String revisionTargetPath = APIConstants.API_REVISION_LOCATION + RegistryConstants.PATH_SEPARATOR +
+                apiUUID +
+                RegistryConstants.PATH_SEPARATOR + revisionId;
+        boolean transactionCommitted = false;
+        Registry registry = null;
+        try {
+            RegistryHolder holder = getRegistry(org.getName());
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:408]: tenat flow is not started

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 9]
```diff
@@ -276,6 +276,168 @@ public PublisherAPI addAPI(Organization org, PublisherAPI publisherAPI) throws A
...
+
+    @Override
+    public void restoreAPIRevision(Organization org, String apiUUID, int revisionId)
+            throws APIPersistenceException {
+
+        boolean transactionCommitted = false;
+        Registry registry = null;
+        APIIdentifier apiId = null;
+        try {
+            RegistryHolder holder = getRegistry(org.getName());
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:350]: tenant flow is not started

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java [hunk 10]
```diff
@@ -276,6 +276,168 @@ public PublisherAPI addAPI(Organization org, PublisherAPI publisherAPI) throws A
...
     }
 
+    @Override
+    public String addAPIRevision(Organization org, String apiUUID, int revisionId) throws APIPersistenceException {
+        String revisionUUID;
+        boolean transactionCommitted = false;
+        Registry registry = null;
+        try {
+            RegistryHolder holder = getRegistry(org.getName());
+            registry = holder.getRegistry();
```
Comment (chamilaadhi) [inline components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:286]: tenant flow not started

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevision.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevision.java:2]: Need to update the year

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevisionDeployment.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright (c) 2020, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevisionDeployment.java:2]: Need to update the year

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java [hunk 13]
```diff
@@ -817,12 +825,22 @@ private static void setMaxTpsFromApiDTOToModel(APIDTO dto, API api) {
...
+    public static APIDTO fromAPItoDTO(API model, APIProvider apiProvider) throws APIManagementException {
 
-        APIProvider apiProvider = RestApiCommonUtil.getLoggedInUserProvider();
+        return fromAPItoDTO(model, false, apiProvider);
+    }
+
+    public static APIDTO fromAPItoDTO(API model, boolean preserveCredentials, APIProvider apiProviderParam)
+            throws APIManagementException {
+        APIProvider apiProvider;
+        if(apiProviderParam != null) {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java:839]: Formatting issue

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java [hunk 14]
```diff
@@ -2759,4 +2801,72 @@ private static JSONObject handleEndpointSecurity(API api, JSONObject endpointSec
...
+        }
+        apiRevisionListDTO.setCount(apiRevisionList.size());
+        apiRevisionListDTO.setList(apiRevisionDTOS);
+        return apiRevisionListDTO;
+    }
+
+    public static APIRevisionDeploymentDTO fromAPIRevisionDeploymenttoDTO(APIRevisionDeployment model) throws APIManagementException {
+        APIRevisionDeploymentDTO apiRevisionDeploymentDTO = new APIRevisionDeploymentDTO();
+        apiRevisionDeploymentDTO.setName(model.getDeployment());
+        if(model.getRevisionUUID() != null) {
```
Comment (mushthaq33) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java:2846]: Formatting issue

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Create/Default/APICreateDefault.jsx [hunk 15]
```diff
@@ -198,45 +201,112 @@ function APICreateDefault(props) {
...
+                    setIsRevisioning(false);
+                    const envList = settings.environment.map((env) => env.name);
+                    const body1 = [];
+                    if (envList.length > 0) {
+                        body1.push({
+                            name: envList[0],
+                            displayOnDevportal: true,
+                        });
+                    }
+                    // for (let i = 0; i < envList.length; i++) {
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Create/Default/APICreateDefault.jsx:227]: Lets remove the commented out code

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Listing/SampleAPI/SampleAPI.jsx [hunk 16]
```diff
@@ -100,6 +105,53 @@ class SampleAPI extends Component {
...
+                        const revisionId = api1.body.id;
+                        const envList = settings.environment.map((env) => env.name);
+                        const body1 = [];
+                        if (envList.length > 0) {
+                            body1.push({
+                                name: envList[0],
+                                displayOnDevportal: true,
+                            });
+                        }
+                        // for (let i = 0; i < envList.length; i++) {
```
Comment (mushthaq33) [inline features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Listing/SampleAPI/SampleAPI.jsx:122]: Remove commented code

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java (hunk 1 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java:1832]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java (hunk 2 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIGatewayManager.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 3 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 4 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java (hunk 5 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 8 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:419]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 9 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:356]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java (hunk 10 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.persistence/src/main/java/org/wso2/carbon/apimgt/persistence/RegistryPersistenceImpl.java:287]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevision.java (hunk 11 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevision.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevisionDeployment.java (hunk 12 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/APIRevisionDeployment.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java (hunk 13 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java]: resolved

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java (hunk 14 above)
Comment (CrowleyRajapakse) [reply components/apimgt/org.wso2.carbon.apimgt.rest.api.publisher.v1.common/src/main/java/org/wso2/carbon/apimgt/rest/api/publisher/v1/common/mappings/APIMappingUtil.java]: resolved

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Create/Default/APICreateDefault.jsx (hunk 15 above)
Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Create/Default/APICreateDefault.jsx]: resolved

File: features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Listing/SampleAPI/SampleAPI.jsx (hunk 16 above)
Comment (CrowleyRajapakse) [reply features/apimgt/org.wso2.carbon.apimgt.publisher.feature/src/main/resources/publisher/source/src/app/components/Apis/Listing/SampleAPI/SampleAPI.jsx]: resolved

**Review by mushthaq33 — APPROVED** [review]

**Pull request merged**